apps/emqx_durable_storage/src/emqx_ds.erl

```diff
@@ -114,6 +123,16 @@
 
 -type get_iterator_result(Iterator) :: {ok, Iterator} | undefined.
 
+%% An opaque term identifying a generation.  Each implementation will possibly add
+%% information to this term to match its inner structure (e.g.: by embedding the shard id,
+%% in the case of `emqx_ds_replication_layer').
+-opaque generation_id() :: term().
```

Comment:
`emqx_ds:list_generations_with_lifetimes/2` takes advantage of being implemented in the same module, so it can construct generation_id()'s.

But any other implementations won't be able to implement the callback with opaque types since opaque terms cannot be constructed outside the module.

Reply:
But it's not constructed in the same module that defines `emqx_ds:generation_id()`, right?  It's analogous to `emqx_ds:iterator()`:  it's opaque too.

Note that here (`emqx_ds`) we define this `generation_id()` to be an opaque term, but the terms themselves are constructed in other modules (`emqx_ds_replication_layer` and `emqx_ds_storage_layer`).

Reply:
I'll change it to mimic the other type more closely, nevertheless.

Reply:
:heavy_check_mark: